pallets/auctions/src/lib.rs

```diff
@@ -0,0 +1,726 @@
+
+            // Add auction to storage
+            Auctions::<T>::insert(
+                nft_id,
+                AuctionData {
+                    creator: creator.clone(),
+                    start_block,
+                    end_block,
+                    start_price,
+                    buy_it_price,
+                    bidders,
+                    marketplace_id,
+                    state: AuctionState::Pending,
+                },
+            );
+
+            // Emit AuctionCreated event
+            Self::deposit_event(Event::AuctionCreated {
+                nft_id,
+                marketplace_id,
+                creator,
+            });
+            // Return a successful DispatchResultWithPostInfo
+            Ok(().into())
+        }
+
+        #[pallet::weight(10_000 + T::DbWeight::get().writes(1))]
+        #[transactional]
+        pub fn cancel_auction(origin: OriginFor<T>, nft_id: NFTId) -> DispatchResultWithPostInfo {
+            let who = ensure_signed(origin)?;
+            let current_block = frame_system::Pallet::<T>::block_number();
+
+            // fetch the data of given nftId
+            let nft_data = T::NFTHandler::get_nft(nft_id).ok_or(Error::<T>::NFTIdInvalid)?;
+
+            // ensure the caller is the owner of NFT
+            ensure!(nft_data.owner == who.clone(), Error::<T>::NftNotOwned);
+
+            let auction = Auctions::<T>::get(nft_id).ok_or(Error::<T>::AuctionDoesNotExist)?;
+
+            // ensure auction has not started
+            ensure!(
+                auction.start_block > current_block,
+                Error::<T>::CannotCancelInProcessAuction
+            );
+
+            // List nft as not for sale
+            T::NFTHandler::set_listed_for_sale(nft_id, false)?;
+
+            // Remove auction from storage
+            Auctions::<T>::remove(nft_id);
+
+            // Emit auction canceled event
+            Self::deposit_event(Event::AuctionCancelled {
+                nft_id,
+                marketplace_id: auction.marketplace_id,
+                creator: who,
+            });
+
+            Ok(().into())
+        }
+
+        #[pallet::weight(10_000 + T::DbWeight::get().writes(1))]
+        #[transactional]
+        pub fn add_bid(
+            origin: OriginFor<T>,
+            nft_id: NFTId,
+            amount: BalanceOf<T>,
+        ) -> DispatchResultWithPostInfo {
+            let who = ensure_signed(origin)?;
+            let current_block = frame_system::Pallet::<T>::block_number();
+
+            // fetch the data of given nftId
+            let nft_data = T::NFTHandler::get_nft(nft_id).ok_or(Error::<T>::NFTIdInvalid)?;
+
+            // fetch data of auction that lists NFT
+            let auction = Auctions::<T>::get(nft_id).ok_or(Error::<T>::AuctionDoesNotExist)?;
+
+            // ensure the nft is in listed for sale state
+            ensure!(
+                nft_data.listed_for_sale == true,
```

Comment:
This is unnecessary. If an auction exists it means that the nft is already listed for sale.